In [1]:
import torch
import numpy as np

# Basic operation

## Judge the type of object

In [5]:
# judge whether an object is a tensor or not.
a = np.array([1, 2, 3, 4])
b = torch.randn(3,3) # generate samples in gaussian distribution.
print("a is a tensor:%s" % torch.is_tensor(a))
print("b is a tensor:%s" % torch.is_tensor(b))


a is a tensor:False
b is a tensor:True


we can also use function:`torch.is_storage(obj)` to judge whether the object is a python storage object.return bool type(true or false)

## Get the element number of tensor 

`torch.numel(input)->int`: return the element number of tensor:input

In [6]:
a = torch.randn(1, 2, 3, 4, 5)
print("the element number of tensor a is: %d" % torch.numel(a))
b = torch.zeros(4, 4)
print("the element number of tensor b is: %d" % torch.numel(b))

the element number of tensor a is: 120
the element number of tensor b is: 16


calculate method: a is 5 dims tensor, total element number is:5! = 120

## Create Ops

### create identity matrix(m * n)

create a inentity matrix:
**`torch.eye(n, m=None, out=None)`**: return a table(2 dims identity matrix) 

**parameters:** 

* n(int):the number of rows.
* m(int, optional):the number of column,if "None":m==n
* out(Tensor, optional): output tensor    

In [9]:
a = torch.eye(3)
print(a)

tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]])


In [10]:
b = torch.eye(2, 3)
print(b)

tensor([[1., 0., 0.],
        [0., 1., 0.]])


### convert numpy object into tensor

function:**`from_numpy(ndarray)`**-> Tensor
Note:
    this function convert a numpy object into Tensor obj.

**Key point:**

* the tensor and numpy object share the same memory space, modify one of them, the other obj will be modified same. 
* the size of returned tensor can't be modified.

In [16]:
a = np.eye(3)
print(a)
t = torch.from_numpy(a)
print(t)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]], dtype=torch.float64)


In [22]:
t[0, 1] = 2
print(a)
# we can see that when we modified tensor, the orginal numpy obj has been modified.
a[0, 0] = 10
print(t)

[[1. 2. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
tensor([[10.,  2.,  0.],
        [ 0.,  1.,  0.],
        [ 0.,  0.,  1.]], dtype=torch.float64)


### ceate tensor object like Arithmetic progression

function:**`torch.linspace(start, end, steps=100, out=None)`**->tensor
return: return a list(1 dim tensor), each element number will be included into the interval:**$[start, end]$** with uniform distribution and the length of list is:`step`.

**parameters:**

* start(float)- start point.
* end(float) - end point.
* step(int) - the generated sample nums between start and end.


In [23]:
torch.linspace(1, 10, steps=10)

tensor([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.])

In [24]:
torch.linspace(-10, 10, steps = 5)

tensor([-10.,  -5.,   0.,   5.,  10.])

function:**`torch.logspace(start, end, steps=100, out=None)`**->tensor.
return: return a list(1 dim tensor), each element number will be included into the interval:**$[10^{start}, 10^{end}]$** with uniform distribution in **logarithmic scale** and the length of list is:`step`.

**parameters:**

* start(float)- start point.
* end(float) - end point.
* step(int) - the generated sample nums between start and end.


In [27]:
torch.logspace(start = -10, end = 10, steps = 20)

tensor([1.0000e-10, 1.1288e-09, 1.2743e-08, 1.4385e-07, 1.6238e-06, 1.8330e-05,
        2.0691e-04, 2.3357e-03, 2.6367e-02, 2.9764e-01, 3.3598e+00, 3.7927e+01,
        4.2813e+02, 4.8329e+03, 5.4556e+04, 6.1585e+05, 6.9519e+06, 7.8476e+07,
        8.8587e+08, 1.0000e+10])

### create a tensor which all elements  are same

function:**torch.ones(*sizes,out=None)** -> tensor. return a tensor which all elements are 1.
**parameters:**

* sizes(int...)  define the shape of tensor.

In [2]:
torch.ones(3,2)

tensor([[1., 1.],
        [1., 1.],
        [1., 1.]])

In [3]:
torch.ones(4)

tensor([1., 1., 1., 1.])

In [9]:
torch.ones(1, 4) == torch.ones(4) # pay attention to the difference of above statements.

tensor([[True, True, True, True]])

### create a tensor with uniform distribution.

function:**torch.rand(*sizes, out=None)**->Tensor. return a tensor which elements follow the uniform distribution sampling from interval $[0, 1)$

**parameters:**

* sizes(int...)  define the shape of tensor.

In [11]:
torch.rand(4,3)

tensor([[0.0276, 0.8242, 0.9889],
        [0.6249, 0.3677, 0.8770],
        [0.4802, 0.4341, 0.6548],
        [0.6123, 0.4190, 0.5652]])

In [12]:
torch.rand(4)

tensor([0.4394, 0.9744, 0.1895, 0.0690])

In [14]:
torch.rand(1,4)

tensor([[0.4931, 0.9895, 0.5173, 0.2326]])

### create a tensor with gaussian distribution.（mean = 0, var = 1）

function:**torch.randn(*sizes, out=None)**->Tensor.return a tensor which elements follow the gaussian distribution.

**parameters:**

* sizes(int...) define the shape of tensor 
* out(Tensor,optinal)

In [2]:
torch.randn(4)

tensor([ 0.1764, -1.3626, -0.7835, -1.5554])

In [3]:
torch.randn(3,4)

tensor([[-0.3917,  0.7055, -0.4928,  0.1336],
        [-0.3742,  0.5645,  0.3063,  0.8041],
        [ 2.5120,  0.0447,  2.5879, -0.1385]])

### create a tensor which elements including random permutation of number.
function:**torch.randperm(n,out=None)**->LongTensor return:Tensor which elements including random permutation from 0 to n-1.


In [5]:
torch.randperm(4)

tensor([0, 3, 2, 1])

In [6]:
torch.randperm(4)

tensor([0, 3, 2, 1])

In [7]:
torch.randperm(10)

tensor([5, 4, 2, 1, 7, 0, 8, 3, 6, 9])

In [8]:
torch.randperm(20)

tensor([ 0, 13, 11, 15,  1, 10,  8,  9, 14, 18, 16,  4, 19,  6, 12,  2,  7,  5,
         3, 17])

### torch.arange(start, end, step=1, out=None) ->Tensor
return an one dimension tensor, $length=(end-start)/step$.

In [9]:
torch.arange(1, 4)

tensor([1, 2, 3])

In [10]:
torch.arange(1, 100, 2)

tensor([ 1,  3,  5,  7,  9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35,
        37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71,
        73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99])

In [11]:
torch.arange(1, 2.5, 0.5)

tensor([1.0000, 1.5000, 2.0000])

### torch.range(start, end, step =1, out=None) -> Tensor
return an one dimension tensor, $length=(end-start)/step$

In [13]:
torch.range(1,4)

/Users/mahongying/anaconda3/envs/torch/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: torch.range is deprecated in favor of torch.arange and will be removed in 0.5. Note that arange generates values in [start; end), not [start; end].
  """Entry point for launching an IPython kernel.


tensor([1., 2., 3., 4.])

In [14]:
torch.arange(1, 4)

tensor([1, 2, 3])

In [15]:
torch.range(1, 4, 0.5)

/Users/mahongying/anaconda3/envs/torch/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: torch.range is deprecated in favor of torch.arange and will be removed in 0.5. Note that arange generates values in [start; end), not [start; end].
  """Entry point for launching an IPython kernel.


tensor([1.0000, 1.5000, 2.0000, 2.5000, 3.0000, 3.5000, 4.0000])

### torch.zeros(*size, out=None)->Tensor

In [16]:
torch.zeros((3,4))

tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [18]:
torch.zeros(4) == torch.zeros((1, 4))

tensor([[True, True, True, True]])

In [19]:
a = torch.rand((3,3))

In [20]:
a

tensor([[0.4965, 0.6675, 0.3276],
        [0.8480, 0.1091, 0.9461],
        [0.4126, 0.3236, 0.7041]])

In [21]:
a + 4

tensor([[4.4965, 4.6675, 4.3276],
        [4.8480, 4.1091, 4.9461],
        [4.4126, 4.3236, 4.7041]])

In [23]:
a + 4 == torch.add(a, 4)

tensor([[True, True, True],
        [True, True, True],
        [True, True, True]])